drivers/counter/counter_mcux_lpc_rtc.c

```diff
@@ -0,0 +1,250 @@
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+	struct mcux_lpc_rtc_data *data = dev->data;
+
+	uint32_t ticks = alarm_cfg->ticks;
+	uint32_t current = mcux_lpc_rtc_read(dev);
+
+	LOG_DBG("Current time is %d ticks", current);
+
+	if (chan_id != 0U) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (data->alarm_callback != NULL) {
+		return -EBUSY;
+	}
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
+		ticks += current;
+	}
+
+	if (ticks < current) {
+		LOG_ERR("Alarm cannot be earlier than current time");
+		return -EINVAL;
+	}
+
+	data->alarm_callback = alarm_cfg->callback;
+	data->alarm_user_data = alarm_cfg->user_data;
+
+	RTC_SetSecondsTimerMatch(config->base, ticks);
+	LOG_DBG("Alarm set to %d ticks", ticks);
+
+	return 0;
+}
+
+static int mcux_lpc_rtc_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	struct mcux_lpc_rtc_data *data = dev->data;
+
+	if (chan_id != 0U) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	data->alarm_callback = NULL;
+
+	return 0;
+}
+
+static int mcux_lpc_rtc_set_top_value(const struct device *dev,
+				  const struct counter_top_cfg *cfg)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+			CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+	struct mcux_lpc_rtc_data *data = dev->data;
+
+	if (cfg->ticks != info->max_top_value) {
+		LOG_ERR("Wrap can only be set to 0x%x.", info->max_top_value);
+		return -ENOTSUP;
+	}
+
+	if (!(cfg->flags & COUNTER_TOP_CFG_DONT_RESET)) {
+		RTC_StopTimer(config->base);
+		RTC_SetSecondsTimerCount(config->base, 0);
+		RTC_StartTimer(config->base);
+	}
+
+	data->top_callback = cfg->callback;
+	data->top_user_data = cfg->user_data;
+
+	return 0;
+}
+
+static uint32_t mcux_lpc_rtc_get_pending_int(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+
+	return RTC_GetStatusFlags(config->base) & RTC_CTRL_ALARM1HZ_MASK;
+}
+
+static uint32_t mcux_lpc_rtc_get_top_value(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+
+	return info->max_top_value;
+}
+
+static void mcux_lpc_rtc_isr(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+	struct mcux_lpc_rtc_data *data = dev->data;
+	counter_alarm_callback_t cb;
+	uint32_t current = mcux_lpc_rtc_read(dev);
+
+
+	LOG_DBG("Current time is %d ticks", current);
+
+	if ((RTC_GetStatusFlags(config->base) & RTC_CTRL_ALARM1HZ_MASK) &&
+	    (data->alarm_callback)) {
+		cb = data->alarm_callback;
+		data->alarm_callback = NULL;
+		cb(dev, 0, current, data->alarm_user_data);
+	}
+
+	if (data->top_callback) {
```

Comment:
it looks a bit strange that top_callback is called always, even when alarm occured? Don't know the RTC details, it just looks strange.